reference-manual/src/main/asciidoc/appendix.adoc

```diff
@@ -283,6 +284,158 @@ The URL used for phone home requests is
 
+[appendix]
+= Migration Guides
+
+This appendix provides guidelines for upgrading to a new Hazelcast Jet
+version. See also the
+https://docs.hazelcast.org/docs/jet/release-notes/index.html[Release Notes]
+document for the changes for each Hazelcast Jet release.
+
+== Upgrading to Hazelcast Jet 4.0
+
+Jet 4.0 is a major version release. According to the semantic versioning
+we apply, this means that in version 4.0 some of the API has changed in
+a breaking way and code written for 3.x may no longer compile against it.
+
+=== Jet on IMDG 4.0
+
+Jet 4.0 uses IMDG 4.0, which is also a major release with its own
+breaking changes. For details see
+https://docs.hazelcast.org/docs/rn/index.html[IMDG Release Notes]
+and
+https://docs.hazelcast.org/docs/latest/manual/html-single/#migration-guides[IMDG Migration Guides].
+
+The most important changes we made and which have affected Jet too are
+as follows:
+
+* We renamed many packages and moved classes around. For details see the
+https://docs.hazelcast.org/docs/rn/index.html#4-0[IMDG Release Notes].
+The most obvious change is that many classes that used to be in the
+general `com.hazelcast.core` package are now in specific packages like
+`com.hazelcast.map` or `com.hazelcast.collection`.
+
+* We moved the serializable variants of `java.util.function` interfaces
+from `com.hazelcast.jet.function` to `com.hazelcast.function`:
+`BiConsumerEx`,`BiFunctionEx`,`BinaryOperatorEx`, `BiPredicateEx`,
+`ComparatorEx`, `ComparatorsEx`, `ConsumerEx`, `FunctionEx`, `Functions`,
+`PredicateEx`, `SupplierEx`, `ToDoubleFunctionEx`, `ToIntFunctionEx`,
+`ToLongFunctionEx`.
+
+* We updated/fixed the type parameters on several methods and classes, for
+example the `EntryProcessor`. See the
+https://docs.hazelcast.org/docs/latest/manual/html-single/#introducing-lambda-friendly-interfaces[relevant section]
+in the IMDG Migration Guide.
+
+* We changed the usage of the term "group" in configuration to "cluster".
+See the code snippet below for an example. This changes a Jet Command Line
+parameter as well (`-g/--groupName` renamed to `-n/--cluster-name`).
++
+[source]
+----
+include::{javasource}/migration/Migration.java[tag=config1]
+----
+
+* We moved `EventJournalConfig` from the top-level `Config` class to
+data structure-specific configs (`MapConfig`,`CacheConfig`).
++
+[source]
+----
+include::{javasource}/migration/Migration.java[tag=config2]
+----
+
+* Async methods that used to return `ICompletableFuture` now return
+`CompletionStage` (see the
+https://docs.hazelcast.org/docs/latest/manual/html-single/#removal-of-icompletablefuture[relevant section]
+in the IMDG Migration Guide).
+
+=== Jet API Changes
+
+We made multiple breaking changes in Jet's own APIs too:
+
+* We removed the Jet-specific interfaces `IMapJet`, `ICacheJet` and
+`IListJet` and now use IMDG's `IMap`, `ICache` and `IList` directly.
+
+* We renamed `Pipeline.drawFrom` to `Pipeline.readFrom` and
+`GeneralStage.drainTo` to `GeneralStage.writeTo`.
```

Comment:
`Pipeline.drawFrom` -> `Pipeline.readTo` and `GeneralStage.drainTo` -> `GeneralStage.writeTo`

Reply:
I think I read this wrong, my mistake..